query-algebrizer/src/validate.rs

```diff
@@ -0,0 +1,234 @@
+            // Each 'leg' must have the same variable set.
+            if or_join.clauses.len() < 2 {
+                Ok(())
+            } else {
+                let mut clauses = or_join.clauses.iter();
+                let template = clauses.next().unwrap().mentioned_variables();
+                for clause in clauses {
+                    if template != clause.mentioned_variables() {
+                        bail!(ErrorKind::NonMatchingVariablesInOrClause);
+                    }
+                }
+                Ok(())
+            }
+        },
+        UnifyVars::Explicit(ref vars) => {
+            // Each leg must use the joined vars.
+            let var_set: BTreeSet<Variable> = vars.iter().cloned().collect();
+            for clause in &or_join.clauses {
+                if !var_set.is_subset(&clause.mentioned_variables()) {
+                    bail!(ErrorKind::NonMatchingVariablesInOrClause);
+                }
+            }
+            Ok(())
+        },
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate mentat_core;
+    extern crate mentat_query;
+    extern crate mentat_query_parser;
+
+    use self::mentat_query::{
+        FindQuery,
+        NamespacedKeyword,
+        OrWhereClause,
+        Pattern,
+        PatternNonValuePlace,
+        PatternValuePlace,
+        PlainSymbol,
+        SrcVar,
+        UnifyVars,
+        Variable,
+        WhereClause,
+    };
+
+    use self::mentat_query_parser::parse_find_string;
+
+    use super::validate_or_join;
+
+    /// Tests that the top-level form is a valid `or`, returning the clauses.
+    fn valid_or_join(parsed: FindQuery, expected_unify: UnifyVars) -> Vec<OrWhereClause> {
+        let mut wheres = parsed.where_clauses.into_iter();
+
+        // There's only one.
+        let clause = wheres.next().unwrap();
+        assert_eq!(None, wheres.next());
+
+        match clause {
+            WhereClause::OrJoin(or_join) => {
+                // It's valid: the variables are the same in each branch.
+                assert_eq!((), validate_or_join(&or_join).unwrap());
+                assert_eq!(expected_unify, or_join.unify_vars);
+                or_join.clauses
+            },
+            _ => panic!(),
+        }
+    }
+
+    /// Test that an `or` is valid if all of its arms refer to the same variables.
+    #[test]
+    fn test_success_or() {
+        let query = r#"[:find [?artist ...]
+                        :where (or [?artist :artist/type :artist.type/group]
+                                   (and [?artist :artist/type :artist.type/person]
+                                        [?artist :artist/gender :artist.gender/female]))]"#;
+        let parsed = parse_find_string(query).unwrap();
+        let clauses = valid_or_join(parsed, UnifyVars::Implicit);
+
+        // Let's do some detailed parse checks.
+        let mut arms = clauses.into_iter();
+        match (arms.next(), arms.next(), arms.next()) {
+            (Some(left),
```

Comment:
nit: one line or different indentation.  (Maybe this is an artifact of GH's layout, but it looks dedented one space.)